### PR TITLE
(doc) Update jenkins scripts

### PR DESCRIPTION
**Pull request by ryanrichter94**
This update fixes a problem with dependencies.

* Updates Get-UpdatedPackage.ps1 
* Updates Update-ProdRepoFromTest.ps1
* Updates ConvertTo-ChocoObject.ps1

### CODE REVIEW

**Comment by adilio** [comment]
Looks like the TOC might have gotten mangled?

File: How-To-Setup-Internal-Package-Repository.md [hunk 1]
```diff
@@ -211,17 +193,18 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
...
       [string]
       $RemoteRepo
+
   )
 
   . .\ConvertTo-ChocoObject.ps1
 
   Write-Verbose "Getting list of local packages from '$LocalRepo'."
-  $localPkgs = choco list --source $LocalRepo | Select-Object -Skip 1 | Select-Object -SkipLast 1 | ConvertTo-ChocoObject
+  $localPkgs = choco list --source $LocalRepo -r | ConvertTo-ChocoObject
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:202]: Can you use the full parameter name here as per the [style guidelines](https://chocolatey.org/docs/commands-reference#scripting-integration-best-practices-style-guide)?

File: How-To-Setup-Internal-Package-Repository.md [hunk 2]
```diff
@@ -211,17 +193,18 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
...
   Write-Verbose "Retrieved list of $(($localPkgs).count) packages from '$Localrepo'."
 
   $localPkgs | ForEach-Object {
       Write-Verbose "Getting remote package information for '$($_.name)'."
-      $remotePkg = choco list $_.name --source $RemoteRepo --exact | Select-Object -Skip 1 | Select-Object -SkipLast 1 | ConvertTo-ChocoObject
+      $remotePkg = choco list $_.name --source $RemoteRepo --exact -r | ConvertTo-ChocoObject
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:207]: Same full parameter name for this one?

File: How-To-Setup-Internal-Package-Repository.md [hunk 3]
```diff
@@ -244,16 +227,18 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
               Write-Verbose "Failed to download package '$($_.name)'"
           }
       }
+
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:230]: Can you remove this blank line between the if and the else?

File: How-To-Setup-Internal-Package-Repository.md [hunk 4]
```diff
@@ -271,35 +256,34 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
   . .\ConvertTo-ChocoObject.ps1
 
   # get all of the packages from the test repo
-  $testPkgs = choco list --source $TestRepo | Select-Object -Skip 1 | Select-Object -SkipLast 1 | ConvertTo-ChocoObject
-  $prodPkgs = choco list --source $ProdRepo | Select-Object -Skip 1 | Select-Object -SkipLast 1 | ConvertTo-ChocoObject
+  $testPkgs = choco list --source $TestRepo -r | ConvertTo-ChocoObject
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:259]: Full parameter name again.

File: How-To-Setup-Internal-Package-Repository.md [hunk 5]
```diff
@@ -271,35 +256,34 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
   . .\ConvertTo-ChocoObject.ps1
 
   # get all of the packages from the test repo
-  $testPkgs = choco list --source $TestRepo | Select-Object -Skip 1 | Select-Object -SkipLast 1 | ConvertTo-ChocoObject
-  $prodPkgs = choco list --source $ProdRepo | Select-Object -Skip 1 | Select-Object -SkipLast 1 | ConvertTo-ChocoObject
+  $testPkgs = choco list --source $TestRepo -r | ConvertTo-ChocoObject
+  $prodPkgs = choco list --source $ProdRepo -r | ConvertTo-ChocoObject
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:260]: Full parameter name again.

File: How-To-Setup-Internal-Package-Repository.md [hunk 6]
```diff
@@ -271,35 +256,34 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
...
-  $testPkgs = choco list --source $TestRepo | Select-Object -Skip 1 | Select-Object -SkipLast 1 | ConvertTo-ChocoObject
-  $prodPkgs = choco list --source $ProdRepo | Select-Object -Skip 1 | Select-Object -SkipLast 1 | ConvertTo-ChocoObject
+  $testPkgs = choco list --source $TestRepo -r | ConvertTo-ChocoObject
+  $prodPkgs = choco list --source $ProdRepo -r | ConvertTo-ChocoObject
   $tempPath = Join-Path -Path $env:TEMP -ChildPath ([GUID]::NewGuid()).GUID
+
   if ($null -eq $testPkgs) {
       Write-Verbose "Test repository appears to be empty. Nothing to push to production."
   }
+
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:266]: Can you remove this blank line.

File: How-To-Setup-Internal-Package-Repository.md [hunk 7]
```diff
@@ -271,35 +256,34 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
...
   elseif ($null -eq $prodPkgs) {
       $pkgs = $testPkgs
   }
+
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:270]: Can you remove this blank line.

File: How-To-Setup-Internal-Package-Repository.md [hunk 8]
```diff
@@ -271,35 +256,34 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
...
   else {
       $pkgs = Compare-Object -ReferenceObject $testpkgs -DifferenceObject $prodpkgs -Property name, version | Where-object SideIndicator -eq '<='
   }
 
   $pkgs | ForEach-Object {
       Write-Verbose "Downloading package '$($_.name)' to '$tempPath'."
-      choco download $_.name --no-progress --output-directory=$tempPath --source=$TestRepo
-
+      choco download $_.name --no-progress --output-directory=$tempPath --source=$TestRepo --ignore-dependencies
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:277]: This actually changes the functionality of the script - why did you add `--ignore-dependencies`?

File: How-To-Setup-Internal-Package-Repository.md [hunk 9]
```diff
@@ -271,35 +256,34 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
...
       if ($LASTEXITCODE -eq 0) {
           $pkgPath = (Get-Item -Path (Join-Path -Path $tempPath -ChildPath '*.nupkg')).FullName
-
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:293]: Can you add this blank lines back in - they help make it visually obvious for the next comments to insert the test code.

File: How-To-Setup-Internal-Package-Repository.md [hunk 10]
```diff
@@ -271,35 +256,34 @@ Jenkins requires several PowerShell scripts to automate the processes. Create a
...
           # #######################
           # INSERT CODE HERE TO TEST YOUR PACKAGE
           # #######################
-
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:297]: Can you add this blank lines back in - they help make it visually obvious for the next comments to insert the test code.

File: How-To-Setup-Internal-Package-Repository.md [hunk 11]
```diff
@@ -325,17 +308,19 @@ Note the section above where you should insert the code to test your packages be
 
 ```powershell
   function ConvertTo-ChocoObject {
+
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:311]: I'm not sure why this blank line was added - can you remove it?

File: How-To-Setup-Internal-Package-Repository.md [hunk 12]
```diff
@@ -325,17 +308,19 @@ Note the section above where you should insert the code to test your packages be
 
 ```powershell
   function ConvertTo-ChocoObject {
+
       [CmdletBinding()]
+
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:313]: I'm not sure why this blank line was added - can you remove it?

File: How-To-Setup-Internal-Package-Repository.md [hunk 13]
```diff
@@ -325,17 +308,19 @@ Note the section above where you should insert the code to test your packages be
...
       [CmdletBinding()]
+
       Param (
           [Parameter(ValueFromPipeline)]
           [string]$InputObject
       )
 
       Process {
-          # format of the 'choco list' output is:
+          # format of the 'choco list -r' output is:
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:320]: Full parameter name again.

File: How-To-Setup-Internal-Package-Repository.md [hunk 14]
```diff
@@ -325,17 +308,19 @@ Note the section above where you should insert the code to test your packages be
...
           # <PACKAGE NAME> <VERSION> (ie. adobereader 2015.6.7)
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:321]: After adding the `-r` parameter the format of the output changes. Can you amend this comment to reflect that?

File: How-To-Setup-Internal-Package-Repository.md [hunk 15]
```diff
@@ -1,24 +1,6 @@
...
...(#internal-package-repositories)auto            - [Install and Configure Chocolatey Server](#install-and-configure-chocolatey-server)auto        - [Install and Configure Jenkins Server](#install-and-configure-jenkins-server)auto            - [Script: `Get-UpdatedPackage.ps1`](#script-get-updatedpackageps1)auto            - [Script: `Update-ProdRepoFromTest.ps1`](#script-update-prodrepofromtestps1)auto            - [Script: `ConvertTo-ChocoObject.ps1`](#script-convertto-chocoobjectps1)auto            - [Create Jenkins Jobs](#create-jenkins-jobs)auto                - [Jenkins Job Details: Update Test Repository](#jenkins-job-details-update-test-repository)auto                - [Jenkins Job Details: Internalize Package](#jenkins-job-details-internalize-package)auto                - [Jenkins Job Details: Update Production Repository](#jenkins-job-details-update-production-repository)auto    - [Test the Jenkins Automation (Exercises)](#test-the-jenkins-automation-exercises)auto        - [Submit a new package](#submit-a-new-package)auto        - [Updating a package from the Chocolatey Community Repository](#updating-a-package-from-the-chocolatey-community-repository)autoauto<!-- /TOC -->
```
Comment (pauby) [inline How-To-Setup-Internal-Package-Repository.md:3]: @adilio has already mentioned this but I wanted to add it here too. The TOC looks like it messed up.

**Comment by pauby** [comment]
I think there is an issue for this?

**Comment by gep13** [comment]
@ryanrichter94 let me know if you have any questions about making some of the suggested amendments to this PR.

File: How-To-Setup-Internal-Package-Repository.md (hunk 8 above)
Comment (steviecoaster) [reply How-To-Setup-Internal-Package-Repository.md:277]: I believe the logic here is that the other scripts will have pulled down and internalized those dependencies, and having `--ignore-dependencies` here will keep choco from re-downloading the potential same dependencies repeatedly, which would require additional error handling/cleanup between packages.

Comment (pauby) [reply How-To-Setup-Internal-Package-Repository.md:277]: If the dependencies for a package ever change then this won't be picked up by anything other than this script so at that stage you potentially have a package either failing to install (as it can't find it's new dependency) or it reached out to chocolatey.org for that dependency. So that needs to be balanced. I appreciate the chances of a package dependency being added and it not being one of the 'common ones' (that is likely to already be internalized) is rare, but it does happen.

The error handling is already in the code (it captures the exit code at line 291 below and uses a Write-Verbose to output the message). Jenkins _shouldn't_ fail the job in this case as there are no terminating errors from the PowerShell. I say _shouldn't_ and I also say _whenever I've used this is demos and talks (which is a lot) I don't remember any problems_ but it should be tested.

File: How-To-Setup-Internal-Package-Repository.md (hunk 15 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 1 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 3 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 4 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 5 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 6 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 9 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md:293]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 10 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 11 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 13 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 14 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 12 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 6 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 2 above)
Comment (ryanrichter94) [reply How-To-Setup-Internal-Package-Repository.md]: Addressed

File: How-To-Setup-Internal-Package-Repository.md (hunk 8 above)
Comment (steviecoaster) [reply How-To-Setup-Internal-Package-Repository.md:292]: I read that code back just now and had my `ah-ha` moment. Ryan has just pushed a commit  to revert this.

File: How-To-Setup-Internal-Package-Repository.md (hunk 15 above)
Comment (ferventcoder) [reply How-To-Setup-Internal-Package-Repository.md:3]: Still showing the change here as weird

You have AlanWalk's TOC code plugin, which requires you to set an EOL in your code settings, as "auto" is what is being inserted here. https://github.com/AlanWalk/markdown-toc/issues/65

Once you've fixed it and pushed again to your repo, it should update these lines of code.

**Comment by pauby** [comment]
Other than the TOC issues I'm happy with this.

**Comment by gep13** [comment]
@pauby @ryanrichter94 @steviecoaster @adilio @ferventcoder I have made a couple changes to this PR.  The first one being the removal of the changes to the TOC which are not required.  I have also updated the Jenkins configuration sections to match exactly what was deployed as part of QDE v1.  The scripts there are known to work as expected, so I think it makes sense to have them match here.

Could I get a final review on the changes in this PR, so that we can get this merged in.  Thanks

@ryanrichter94 I have had to force push the changes onto your branch, so this will now be out of sync with what is now on the branch that is in GitHub.  Let me know if you have any questions about getting those two things synced up again.

**Comment by adilio** [comment]
LGTM 🚀

**Comment by gep13** [comment]
@ryanrichter94 thanks for doing the initial work in getting this updated, appreciate it!
